src/mon/Elector.cc

```diff
@@ -381,10 +418,101 @@ void Elector::handle_nak(MonOpRequestRef op)
   // the end!
 }
 
+void Elector::begin_peer_ping(int peer)
+{
+  if (live_pinging.count(peer)) {
+    return;
+  }
+
+  dout(5) << __func__ << " against " << peer << dendl;
+
+  peer_tracker.report_live_connection(peer, 0); // init this peer as existing
+  live_pinging.insert(peer);
+  peer_acked_ping[peer] = ceph_clock_now();
+  send_peer_ping(peer);
+  mon->timer.add_event_after(ping_timeout / PING_DIVISOR,
+			     new C_MonContext{mon, [this, peer](int) {
+				 ping_check(peer);
+			       }});
+}
+
+void Elector::send_peer_ping(int peer, const utime_t *n)
+{
+  dout(10) << __func__ << " to peer " << peer << dendl;
+
+  utime_t now;
+  if (n != NULL) {
+    now = *n;
+  } else {
+    now = ceph_clock_now();
```

Comment:
Yeah I'll switch to that

Reply:
Hmm actually this quickly spiraled into a mess since we send the stamps over the wire so I don't think so. (We don't seem to have encoders for the time_point types?)

Reply:
it's due to the mono clock variants not having the requisite {from_,to_}timespec methods, from 427655e39180640fb1de56cef13c4ed8c34fa3bb - looks like the idea is to make sure we don't get a non-monotonic value from the network/disk - @cbodley is that correct? would it make sense to use mono clock locally and convert it when sending over the network? how would you approach this?

Reply:
mono_clock is good for measuring elapsed time on one machine. but when you're comparing timestamps between machines, you do want the real_clock. if you're concerned about the cost of sampling (and don't need lots of precision), you might consider `ceph::coarse_real_clock::now()`

cc @adamemerson

Reply:
The CLOCK_MONOTONIC timers give a time relative to some local timepoint (on Linux it's a time shortly after boot) so there's no way for one machine to use another's monotonic time_point.

You can share durations or real times, however.